beacon-chain/p2p/peers/score_block_providers.go

```diff
@@ -0,0 +1,278 @@
+	ReturnedBlocksWeight float64
+	// SlowReturnedBlocksPenalty defines a penalty applied to score, if blocks were requested,
+	// but none have been returned yet (to distinguish between non-responsive peers and peers that
+	// haven't been requested any blocks yet).
+	SlowReturnedBlocksPenalty float64
+	// ProcessedBlocksWeight defines weight of a processed/requested ratio in overall score.
+	ProcessedBlocksWeight float64
+	// SlowProcessedBlocksPenalty defines a penalty applied to score, if blocks have been
+	// requested and returned, but none have been processed yet. Allows distinguishing between
+	// peers that haven't yet returned anything and peers that returned garbage.
+	SlowProcessedBlocksPenalty float64
+	// DecayInterval defines how often requested/returned/processed stats should be decayed.
+	DecayInterval time.Duration
+	// Decay specifies the factor (must be < 1.0) by which block provider's stats is decayed.
+	Decay float64
+}
+
+// newBlockProviderScorer creates block provider scoring service.
+func newBlockProviderScorer(
+	ctx context.Context, store *peerDataStore, config *BlockProviderScorerConfig) *BlockProviderScorer {
+	if config == nil {
+		config = &BlockProviderScorerConfig{}
+	}
+	scorer := &BlockProviderScorer{
+		ctx:    ctx,
+		config: config,
+		store:  store,
+	}
+	if scorer.config.StartScore == 0.0 {
+		scorer.config.StartScore = DefaultBlockProviderStartScore
+	}
+	if scorer.config.ReturnedBlocksWeight == 0.0 {
+		scorer.config.ReturnedBlocksWeight = DefaultBlockProviderReturnedBlocksWeight
+	}
+	if scorer.config.SlowReturnedBlocksPenalty == 0.0 {
+		scorer.config.SlowReturnedBlocksPenalty = DefaultSlowReturnedBlocksPenalty
+	}
+	if scorer.config.ProcessedBlocksWeight == 0.0 {
+		scorer.config.ProcessedBlocksWeight = DefaultBlockProviderProcessedBlocksWeight
+	}
+	if scorer.config.SlowProcessedBlocksPenalty == 0.0 {
+		scorer.config.SlowProcessedBlocksPenalty = DefaultSlowProcessedBlocksPenalty
+	}
+	if scorer.config.DecayInterval == 0 {
+		scorer.config.DecayInterval = DefaultBlockProviderDecayInterval
+	}
+	if scorer.config.Decay == 0.0 {
+		scorer.config.Decay = DefaultBlockProviderDecay
+	}
+	return scorer
+}
+
+// Score calculates and returns total score based on returned and processed blocks.
+func (s *BlockProviderScorer) Score(pid peer.ID) float64 {
+	s.store.RLock()
+	defer s.store.RUnlock()
+	return s.score(pid)
+}
+
+// score is a lock-free version of Score.
+func (s *BlockProviderScorer) score(pid peer.ID) float64 {
+	score := s.Params().StartScore
+	peerData, ok := s.store.peers[pid]
+	if ok && peerData.requestedBlocks > 0 {
+		// Score returned/requested ratio.
+		returnedBlocksScore := float64(peerData.returnedBlocks) / float64(peerData.requestedBlocks)
+		returnedBlocksScore = returnedBlocksScore * s.config.ReturnedBlocksWeight
+		score += returnedBlocksScore
+
+		if peerData.requestedBlocks > peerData.returnedBlocks {
```

Comment:
same here, this doesnt take into account that there are skipped slots. Peers could be unfairly penalised.